framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java

```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
+    private TreeSet<CheckerMessage> messageStore = null;
+
+    /**
+     * If this is a compound checker or a subchecker of a compound checker, then the message is
+     * stored until all message from all checkers for the compilation unit are issued.
+     *
+     * <p>Otherwise, it prints the message.
+     */
+    @Override
+    protected void printMessage(
+            Diagnostic.Kind kind, String message, Tree source, CompilationUnitTree root) {
+        assert this.currentRoot == root;
+        if (messageStore == null) {
+            super.printMessage(kind, message, source, root);
+        } else {
+            CheckerMessage checkerMessage = new CheckerMessage(kind, message, source, this);
+            messageStore.add(checkerMessage);
+        }
+    }
+
+    /**
+     * Prints error messages for this checker and all subcheckers such that the errors are ordered
+     * by line and column number and then by checker. (See checkerMessageComparator for more precise
+     * order.)
+     *
+     * @param unit Current compilation unit
+     */
+    private void printCollectedMessages(CompilationUnitTree unit) {
+        if (messageStore != null) {
+            for (CheckerMessage pos : messageStore) {
+                super.printMessage(pos.kind, pos.message, pos.source, unit);
+            }
+        }
+    }
+
+    /**
+     * Sort by position at which the error will be printed, then by the order in which the checkers
+     * run, then by kind of message, and finally by the message string.
+     */
+    private final Comparator<CheckerMessage> checkerMessageComparator =
+            new Comparator<CheckerMessage>() {
+
+                @Override
+                public int compare(CheckerMessage o1, CheckerMessage o2) {
+                    int byPos = InternalUtils.compareDiagnosticPosition(o1.source, o2.source);
+                    if (byPos != 0) {
+                        return byPos;
+                    }
+
+                    int o1Checker = BaseTypeChecker.this.getSubcheckers().indexOf(o1.checker);
+                    int o2Checker = BaseTypeChecker.this.getSubcheckers().indexOf(o2.checker);
+                    if (o1Checker != o2Checker) {
+                        if (o1Checker == -1) {
+                            o1Checker = BaseTypeChecker.this.getSubcheckers().size();
+                        }
+                        if (o2Checker == -1) {
+                            o2Checker = BaseTypeChecker.this.getSubcheckers().size();
+                        }
+                        return Integer.compare(o1Checker, o2Checker);
+                    }
+
+                    int kind = o1.kind.compareTo(o2.kind);
+                    if (kind != 0) {
+                        return kind;
+                    }
+
+                    return o1.message.compareTo(o2.message);
+                }
+            };
+
+    /** Represents a message (e.g., an error message) issued by a checker. */
+    private class CheckerMessage {
```

Comment:
Can this class be static - which would safe us from storing a reference to the enclosing class?